packages/runtime/container-runtime/src/runWhileConnectedCoordinator.ts

```diff
@@ -3,54 +3,84 @@
+    stop(reason: SummarizerStopReason): void;
 }
 
 /**
- * Start result indicating that the start was not successful.
+ * Can be useful in testing as well as in places where caller does not use cancellation.
+ * This object implements ISummaryCancellationToken interface but cancellation is never leveraged.
  */
-export interface INotStartedResult {
-    started: false;
-    message: "DisconnectedBeforeRun" | "NeverConnectedBeforeRun";
-}
-
-export interface IConnectableRuntime {
-    readonly connected: boolean;
-    readonly clientId: string | undefined;
-    once(event: "connected", listener: () => void): this;
-    on(event: "disconnected", listener: () => void): this;
-}
+export const neverCancelledToken: ISummaryCancellationToken = {
+    cancelled: false,
+    waitCancelled: new Promise(() => {}),
+};
 
 /**
  * Helper class to coordinate something that needs to run only while connected.
  * This provides promises that resolve as it starts or stops.  Stopping happens
  * when disconnected or if stop() is called.
  */
-export class RunWhileConnectedCoordinator {
+export class RunWhileConnectedCoordinator implements ICancellableSummarizerController {
     private everConnected = false;
-    private readonly stopDeferred = new Deferred<void>();
+    private _cancelled = false;
+    private readonly stopDeferred = new Deferred<SummarizerStopReason>();
+
+    public get cancelled() {
+        if (!this._cancelled) {
+            assert(this.runtime.deltaManager.active, "We should never connect as 'read'");
+
+            // This check can't be enabled in current design due to lastSummary flow, where
+            // summarizer for closed container stays around and can produce one more summary.
+            // Currently we solve the problem of overlapping summarizer by doing wait in
+            // SummaryManager.createSummarizer()
+            // Better solution would involve these steps:
+            // 1. Summarizer selection logic should chose summarizing client (with clientType === "summarizer")
+            // if such client exists.
+            // 2. Summarizer should be updated about such changes (to update onBehalfOfClientId)
+            //
+            // assert(this.runtime.summarizerClientId === this.onBehalfOfClientId ||
+            //    this.runtime.summarizerClientId === this.runtime.clientId, "onBehalfOfClientId");
+        }
+
+        return this._cancelled;
+    }
+
+    /**
+     * Returns a promise that resolves once stopped either externally or by disconnect.
+     */
+     public get waitCancelled(): Promise<SummarizerStopReason> {
+        return this.stopDeferred.promise;
+    }
 
     public constructor(private readonly runtime: IConnectableRuntime) {
         // Try to determine if the runtime has ever been connected
         if (this.runtime.connected) {
             this.everConnected = true;
-        } else {
+        } else if (this.runtime.disposed) {
+            this.stop("summarizeClientDisconnected");
```

Comment:
nit: summarizeClientDisconnected -> summarizerClientDisconnected.

Reply:
I noticed that too :) I'll rename in separate PR not to pile more stuff into already big PR.